std/algorithm/comparison.d

```diff
@@ -732,78 +731,133 @@ template equal(alias pred = "a == b")
-        }
-        // if one of the arguments is a string and the other isn't, then auto-decoding
-        // can be avoided if they have the same ElementEncodingType
-        else static if (is(typeof(pred) == string) && pred == "a == b" &&
-            isAutodecodableString!Range1 != isAutodecodableString!Range2 &&
-            is(ElementEncodingType!Range1 == ElementEncodingType!Range2))
+        enum isEquableToR(T) = is(typeof({ return R.init == T.init; }));
+
+        // start by detecting default pred and compatible dynamic array
+        static if (is(typeof(pred) == string) &&
+                   pred == "a == b" &&
+                   isArray!R &&
+                   allSatisfy!(isArray, Ss) &&
+                   allSatisfy!(isEquableToR, Ss))
         {
-            import std.utf : byCodeUnit;
-
-            static if (isAutodecodableString!Range1)
-            {
-                return equal(r1.byCodeUnit, r2);
-            }
-            else
+            foreach (ref s; ss)
             {
-                return equal(r2.byCodeUnit, r1);
+                if (r != s) return false;
             }
+            return true;
         }
-        //Try a fast implementation when the ranges have comparable lengths
-        else static if (hasLength!Range1 && hasLength!Range2 && is(typeof(r1.length == r2.length)))
+        // use fast implementation when the ranges have comparable lengths
+        else static if (allSatisfy!(hasLength, R, Ss))
         {
-            immutable len1 = r1.length;
-            immutable len2 = r2.length;
-            if (len1 != len2) return false; //Short circuit return
-
-            //Lengths are the same, so we need to do an actual comparison
-            //Good news is we can squeeze out a bit of performance by not checking if r2 is empty
-            for (; !r1.empty; r1.popFront(), r2.popFront())
+            // check equal lengths
+            foreach (ref s; ss)
+            {
+                if (r.length != s.length) { return false; }
+            }
+            // check equal contents
+            for (; !r.empty; r.popFront())  // for each element in first range `r`
             {
-                if (!binaryFun!(pred)(r1.front, r2.front)) return false;
+                foreach (ref s; ss) // for each other range `s`
+                {
+                    if (!binaryFun!(pred)(r.front, s.front)) return false;
+                    s.popFront();
+                }
             }
             return true;
         }
+        // generic case, we have to walk all ranges making sure neither is empty
         else
         {
-            //Generic case, we have to walk both ranges making sure neither is empty
-            for (; !r1.empty; r1.popFront(), r2.popFront())
+            for (; !r.empty; r.popFront())  // for each element in first range `r`
+            {
+                foreach (ref s; ss)
+                {
+                    static if (!isInfinite!(typeof(s)))
```

Comment:
No need for this `isInfinite` business, just use `empty` (it will optimize away code for infinite ranges).